recon/src/recon.rs

```diff
@@ -429,23 +428,97 @@ where
     }
 }
 
+/// Represents reasons the store is unwilling to persist keys and values
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum InvalidItem<K: Key> {
+    /// The key or event data could not be parsed
+    InvalidFormat {
+        /// The key associated with the invalid data
+        key: K,
+    },
+    /// The data did not have a valid signature
+    InvalidSignature {
+        /// The key associated with the invalid data
+        key: K,
+    },
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// Represents items the store may be able to persist in the future
+pub struct PendingItem<K>
+where
+    K: Key,
+{
+    /// The key this item needs in order to be processed.
+    pub required_key: K,
+    /// The item that could not be stored. Can be retried in the future if
+    /// the `required_key` is discovered.
+    pub item: ReconItem<K>,
+}
+
+impl<K: Key> PendingItem<K> {
+    /// Construct a pending item
+    pub fn new(required_key: K, item: ReconItem<K>) -> Self {
+        Self { required_key, item }
+    }
+}
+
 /// The result of an insert operation.
-#[derive(Clone, Debug, PartialEq, Eq, Hash, Default)]
-pub struct InsertResult {
-    /// A true/false list indicating whether or not the key was new.
-    /// It is in the same order as the input list of keys.
-    pub keys: Vec<bool>,
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+pub struct InsertBatch<K>
```

Comment:
nitpick, when I see this type name I assume it contains the list of entries to be inserted.  I actually think the original name of `InsertResult` is clearer to me.  Don't feel that strongly though

Reply:
I'm fine leaving it `InsertResult`.. I changed it because Result<InsertResult> is a bit odd (often having Result in the name implies it's a type alias of Result<T,E> for the crate). I originally intended to include a batch ID and maybe a channel but didn't go that route due to complexity managing the state over the connection.

Reply:
Hmm yeah, the convention around the word Result is fair... `InsertOutcome`?  `InsertStatus`?  `InsertDetails`?

Reply:
I think any of those would be good... but it is _kind of_ a result since it contains info about "invalid" things.. so I just left it for now. We can rename it if we feel it's important in a future PR.